### ISSUE TITLE
HTCondor was never updated with environment support, causing "ValueError: too many values to unpack (expected 6)", and also can't handle the HTCondor schedd being busy

### ISSUE THREAD

**Issue by amizeranschi** [DataBiosphere/toil#4363]
Following up from https://github.com/DataBiosphere/toil/issues/4315, I managed to successfully submit a Cactus job to HTCondor, using the following job descriptor:

```
Executable = /data/share/cactus_conda/bin/cactus
Arguments = " ./jobStore genomes.txt genomes.hal --logFile cactus.log --realTimeLogging --batchSystem htcondor --consCores 32 --disableCaching "
Log = cactus.$(process).log
Output = cactus.$(process).out
Error = cactus.$(process).err
request_cpus = 1
Queue 1
```

However, this failed with the errors from the title. The relevant output is below:

```
[2023-02-04T08:26:06+0100] [MainThread] [W] [toil.common] Batch system does not support auto-deployment. The user script ModuleDescriptor(dirPath='/data/share/cactus_conda/lib/python3.10/site-packages', name='cactus.progressive.cactus_progressive', fromVirtualEnv=False) will have to be present at the same location on every worker.
[2023-02-04T08:26:06+0100] [MainThread] [I] [toil.job] Saving graph of 1 jobs, 1 new
[2023-02-04T08:26:06+0100] [MainThread] [I] [toil.job] Processing job 'progressive_workflow' kind-progressive_workflow/instance-fgnhd28s v0
[2023-02-04T08:26:06+0100] [MainThread] [I] [toil] Running Toil version 5.9.0-8155e0a981f4d728762a7cbc920b7ed544fe4ae7 on host exec-02.novalocal.
[2023-02-04T08:26:06+0100] [MainThread] [I] [toil.realtimeLogger] Starting real-time logging.
[2023-02-04T08:26:06+0100] [MainThread] [I] [toil.leader] Issued job 'progressive_workflow' kind-progressive_workflow/instance-fgnhd28s v1 with job batch system ID: 0 and disk: 2.0 Gi, memory: 2.0 Gi, cores: 1, accelerators: [], preemptible: False
[2023-02-04T08:26:07+0100] [Thread-2  ] [E] [toil.batchSystems.abstractGridEngineBatchSystem] GridEngine like batch system failure
Traceback (most recent call last):
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/abstractGridEngineBatchSystem.py", line 257, in run
    while self._runStep():
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/abstractGridEngineBatchSystem.py", line 246, in _runStep
    activity |= self.createJobs(newJob)
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/htcondor.py", line 45, in createJobs
    jobID, cpu, memory, disk, jobName, command = self.waitingJobs.pop(0)
ValueError: too many values to unpack (expected 6)
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/data/share/cactus_conda/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/abstractGridEngineBatchSystem.py", line 257, in run
    while self._runStep():
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/abstractGridEngineBatchSystem.py", line 246, in _runStep
    activity |= self.createJobs(newJob)
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/htcondor.py", line 45, in createJobs
    jobID, cpu, memory, disk, jobName, command = self.waitingJobs.pop(0)
ValueError: too many values to unpack (expected 6)
/data/share/cactus_conda/lib/python3.10/site-packages/htcondor/_deprecation.py:119: FutureWarning: The "requirements" argument of xquery() will be renamed to "constraint" in a future release. For now, it will accept both names.
  warnings.warn(
[2023-02-04T08:26:08+0100] [MainThread] [I] [toil.leader] 0 jobs are running, 1 jobs are issued and waiting to run
[2023-02-04T09:26:09+0100] [MainThread] [I] [toil.leader] 0 jobs are running, 1 jobs are issued and waiting to run
```

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1284)
┆Issue Number: TOIL-1284

**Comment by amizeranschi** [comment]
An update: Looks like the error comes from here: https://github.com/DataBiosphere/toil/blob/master/src/toil/batchSystems/htcondor.py#L45

I added the following lines before the line mentioned above:

```
                temp = self.waitingJobs.pop(0)
                logger.warning("Type of temp: %s", type(temp))
                logger.warning("Length of temp: %r", len(temp))
                logger.warning("Value of temp: %s", temp)
```

and this was the output:

```
[2023-02-07T12:20:51+0100] [Thread-2  ] [W] [toil.batchSystems.htcondor] Type of temp: <class 'tuple'>
[2023-02-07T12:20:51+0100] [Thread-2  ] [W] [toil.batchSystems.htcondor] Length of temp: 7
[2023-02-07T12:20:51+0100] [Thread-2  ] [W] [toil.batchSystems.htcondor] Value of temp: (0, 1, 2147483648, 2147483648, 'progressive_workflow', '_toil_worker progressive_workflow file:/data/share/Cipri-Ilx-NGS/de-novo/cactus-RO-POL-ASP-PAR__STAR_HTCondor/jobStore kind-progressive_workflow/instance-buwtqt6i --context gASVzgAAAAAAAACMIXRvaWwuYmF0Y2hTeXN0ZW1zLmNsZWFudXBfc3VwcG9ydJSMFFdvcmtlckNsZWFudXBDb250ZXh0lJOUKYGUfZSMEXdvcmtlckNsZWFudXBJbmZvlIwldG9pbC5iYXRjaFN5c3RlbXMuYWJzdHJhY3RCYXRjaFN5c3RlbZSMEVdvcmtlckNsZWFudXBJbmZvlJOUKE5OjCRiMzlmNmY0Yi03Y2U3LTRlZmMtOTE3Ni1hZmY2MmZkZjhmMzSUjAZhbHdheXOUdJSBlHNiLg==', {'OMP_NUM_THREADS': '1'})
```

It looks like after the command string there is an unexpected 7th element with the value `{'OMP_NUM_THREADS': '1'}`.

**Comment by amizeranschi** [comment]
Another update: simply including a seventh variable e.g. `omp_threads` into the list that is [being created here](https://github.com/DataBiosphere/toil/blob/master/src/toil/batchSystems/htcondor.py#L45) got rid of the problem mentioned above.

Now, the job crashes due to a new error: `htcondor.HTCondorIOError: Failed to connect to schedd`. A bit of googling revealed an older discussion that seems to imply this issue is related to multithreading:

https://www-auth.cs.wisc.edu/lists/htcondor-users/2018-January/msg00118.shtml

This issue is on a level above what I'm confidently able to deal with. @adamnovak or @Hexotical could you offer any suggestions on how to deal with this?

Below is the full stack trace after I enabled debug logs in Toil:

```
[2023-02-08T08:46:44+0100] [MainThread] [I] [toil.leader] Issued job 'progressive_workflow' kind-progressive_workflow/instance-_704rlh6 v1 with job batch system ID: 0 and disk: 2.0 Gi, memory: 2.0 Gi, cores: 1, accelerators: [], preemptible: False
[2023-02-08T08:46:44+0100] [MainThread] [D] [toil.bus] Notifying toil.bus.JobIssuedMessage with message: JobIssuedMessage(job_type='progressive_workflow', job_id='kind-progressive_workflow/instance-_704rlh6', toil_batch_id=0)
[2023-02-08T08:46:44+0100] [MainThread] [D] [toil.bus] Notifying toil.bus.QueueSizeMessage with message: QueueSizeMessage(queue_size=1)
[2023-02-08T08:46:45+0100] [Thread-2  ] [D] [toil.batchSystems.abstractGridEngineBatchSystem] No activity, sleeping for 1s
[2023-02-08T08:46:45+0100] [Thread-2  ] [D] [toil.batchSystems.htcondor] Submitting 'executable = /bin/sh\ntransfer_executable = False\narguments = "-c \'_toil_worker progressive_workflow file:/data/share/Cipri-Ilx-NGS/de-novo/cactus-RO-POL-ASP-PAR__STAR_HTCondor/jobStore kind-progressive_workflow/instance-_704rlh6 --context gASVzgAAAAAAAACMIXRvaWwuYmF0Y2hTeXN0ZW1zLmNsZWFudXBfc3VwcG9ydJSMFFdvcmtlckNsZWFudXBDb250ZXh0lJOUKYGUfZSMEXdvcmtlckNsZWFudXBJbmZvlIwldG9pbC5iYXRjaFN5c3RlbXMuYWJzdHJhY3RCYXRjaFN5c3RlbZSMEVdvcmtlckNsZWFudXBJbmZvlJOUKE5OjCQyMjM4MjEyZC1jMzQ5LTQ1ZjEtYTQxNy0xZDg1ZmZkOTk5ZjeUjAZhbHdheXOUdJSBlHNiLg==\'"\nenvironment = "TOIL_RT_LOGGING_ADDRESS=\'10.252.254.166:2836\' TOIL_RT_LOGGING_LEVEL=\'Debug\'"\ngetenv = True\nshould_transfer_files = Yes\noutput = /var/lib/condor/execute/dir_34978/toil_2238212d-c349-45f1-a417-1d85ffd999f7.0.$(cluster).out.log\nerror = /var/lib/condor/execute/dir_34978/toil_2238212d-c349-45f1-a417-1d85ffd999f7.0.$(cluster).err.log\nlog = /var/lib/condor/execute/dir_34978/toil_2238212d-c349-45f1-a417-1d85ffd999f7.0.$(cluster).events.log\nrequest_cpus = 1\nrequest_memory = 2097152.000KB\nrequest_disk = 2097152.000KB\nleave_in_queue = (JobStatus == 4)\nMY.IsToilJob = True\nMY.ToilJobID = 0\nMY.ToilJobName = "progressive_workflow"\nMY.ToilJobKilled = False\n'
[2023-02-08T08:46:45+0100] [Thread-2  ] [D] [toil.batchSystems.htcondor] Connecting to HTCondor Schedd on local machine
[2023-02-08T08:46:46+0100] [MainThread] [D] [toil.batchSystems.htcondor] Connecting to HTCondor Schedd on local machine
/data/share/cactus_conda/lib/python3.10/site-packages/htcondor/_deprecation.py:119: FutureWarning: The "requirements" argument of xquery() will be renamed to "constraint" in a future release. For now, it will accept both names.
  warnings.warn(
[2023-02-08T08:46:50+0100] [Thread-2  ] [E] [toil.batchSystems.abstractGridEngineBatchSystem] GridEngine like batch system failure
Traceback (most recent call last):
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/abstractGridEngineBatchSystem.py", line 257, in run
    while self._runStep():
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/abstractGridEngineBatchSystem.py", line 246, in _runStep
    activity |= self.createJobs(newJob)
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/htcondor.py", line 52, in createJobs
    batchJobID = self.submitJob(submitObj)
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/htcondor.py", line 123, in submitJob
    with schedd.transaction() as txn:
  File "/data/share/cactus_conda/lib/python3.10/site-packages/htcondor/_lock.py", line 70, in wrapper
    rv = func(*args, **kwargs)
htcondor.HTCondorIOError: Failed to connect to schedd.
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/data/share/cactus_conda/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/abstractGridEngineBatchSystem.py", line 257, in run
    while self._runStep():
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/abstractGridEngineBatchSystem.py", line 246, in _runStep
    activity |= self.createJobs(newJob)
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/htcondor.py", line 52, in createJobs
    batchJobID = self.submitJob(submitObj)
  File "/data/share/cactus_conda/lib/python3.10/site-packages/toil/batchSystems/htcondor.py", line 123, in submitJob
    with schedd.transaction() as txn:
  File "/data/share/cactus_conda/lib/python3.10/site-packages/htcondor/_lock.py", line 70, in wrapper
[2023-02-08T08:46:50+0100] [MainThread] [I] [toil.leader] 0 jobs are running, 1 jobs are issued and waiting to run
    rv = func(*args, **kwargs)
htcondor.HTCondorIOError: Failed to connect to schedd.
[2023-02-08T08:47:16+0100] [MainThread] [D] [toil.batchSystems.htcondor] Connecting to HTCondor Schedd on local machine
```

**Comment by adamnovak** [comment]
The problem with the new last item looks to be that, in https://github.com/DataBiosphere/toil/commit/81ca29c797bb802e57088b02717cf6695d9bdc2f#diff-0f2719dd6171193d6a36d8c3f0ec35139002dfc70cd665f9e3e6b83d3ec0ded2R110, we changed the type of the argument to `createJobs`, but we never made the corresponding change in the HTCondor batch system code. And either we have turned off type checking for that file, or the `Any` type in the argument means we didn't catch that this change breaks the HTCondor batch system.

As for the threading, it does indeed look like we have 2 threads that want to talk to htcondor at the same time. It looks like if one thread is doing something, the connection from the other thread might fail, which we don't handle.

But I'm not sure that that that is really the cause of the HTCondorIOError here; it looks like *neither* thread is managing to even connect, and I think we only have 2 threads here (the main one and the one shipping new jobs to HTCondor). Maybe doing some backoff or locking around the connect calls would help anyway?

**Comment by amizeranschi** [comment]
Hi @adamnovak, thanks for looking into this. I tried [the following changes](https://github.com/DataBiosphere/toil/compare/master...amizeranschi:toil:master), but these had no effect on the error mentioned above (`Failed to connect to schedd`).

**Comment by amizeranschi** [comment]
I've reinstalled all of my testing environment from scratch, together with your latest Toil enhancements. Then I tried the Hello World python script that you've linked [here](https://github.com/DataBiosphere/toil/pull/4396#issuecomment-1468589028) and it worked, using HTCondor's Python module v. 8.9.13 and HTCondor native v. 8.8.15.

I've then upgraded the Python htcondor module to 10.3.0 and this also worked. 

When testing my Cactus job, however, I ran into the same trouble as before (failing to connect to schedd). This suggests that either there's something wrong with Cactus itself, or with the way I was running Cactus through htcondor. The way I was running it was through a job description file (cactus.job) that looks like this:

```
Executable = /data/share/cactus_conda/bin/cactus
Arguments = \" ./jobStore genomes.txt genomes.hal --logDebug --logFile cactus.log --realTimeLogging --batchSystem htcondor --consCores 32 --disableCaching \"
Log = cactus.\$(process).log
Output = cactus.\$(process).out
Error = cactus.\$(process).err
request_cpus = 1
Queue 1
```

Which I then submitted via `condor_submit cactus.job`. 

Notably, running Cactus directly via the following command works for me:

```
/data/share/cactus_conda/bin/cactus ./jobStore genomes.txt genomes.hal --logDebug --logFile cactus.log --realTimeLogging --batchSystem htcondor --consCores 32 --disableCaching 
```

No idea why it wouldn't work from a job description file, though.

@adamnovak The reason I went via the job descriptor file route (and never considered running Cactus directly) is that I initially wanted to run Cactus via Docker. In that case, as far as I can see, the only way to run a Docker image on HTCondor is by using a job descriptor file, as shown here: https://htcondor.readthedocs.io/en/latest/users-manual/docker-universe-applications.html

I will test this as soon as there will be an updated Cactus Docker image with an updated Toil (which includes these recent changes). However, I am worried that the same "failed to connect to schedd" errorrs wil reappear in the Docker scenario.

**Comment by adamnovak** [comment]
@amizeranschi I think your problem is probably that, when you run Cactus through the job description file, it is no longer running on the head node; it ends up running the main Cactus leader program on one of the worker nodes of the cluster. That's fine, except that then when the Cactus run says:

```
[2023-03-13T20:27:40+0100] [Thread-2  ] [D] [toil.batchSystems.htcondor] Connecting to HTCondor Schedd on local machine
```

then it is actually saying it is doing the wrong thing, because HTCondor's `schedd` is *not* on the local machine from the Cactus leader process's perspective. It is on your original interactive login machine instead.

`TOIL_HTCONDOR_SCHEDD` is still not in the docs, but if you set it in the Cactus leader process's environment to the address of the HTCondor `schedd` (something like the IP address of the head node, or the IP address of the head node, a colon, and the port that `schedd` listens on), then the Cactus leader will try to talk to the real head node of the cluster and not the nonexistent HTCondor head node install on the worker machine where it is running.

Is there some way that your particular cluster exposes information about the head node to jobs running on the worker nodes? If you wanted to use a job descriptor file to submit a job that was a Bash script that then turned around and submitted more jobs, how would you do it? Maybe we can get Toil to pick up the right HTCondor head node information automatically somehow.

**Comment by amizeranschi** [comment]
I tried setting `TOIL_HTCONDOR_SCHEDD` on the central manager node, where I submit the jobs, but this didn't get rid of the failed to connect to schedd errors when submitting the job via the job description file. I tried setting `TOIL_HTCONDOR_SCHEDD` to the manager node's IP first, then to IP:port.

I remember running into a similar issue some years ago when attempting to run bcbio-nextgen on an SGE cluster. The cluster manager process from bcbio ended up running on a worker node, where it was unable to spawn further jobs and the whole thing would hang. The way we solved it then was to tie that original job submission (which spawned the cluster manager) to a specific worker node, which was set up to also act as a job submit node. 

I will look into how this can be done with htcondor. Their docs mention that this is definitely possible, i.e. that any node can be set up to allow job submission: https://htcondor.readthedocs.io/en/latest/admin-manual/introduction-admin-manual.html#the-different-roles-a-machine-can-play

**Comment by amizeranschi** [comment]
BTW @adamnovak when running the previous successful Cactus job via htcondor (where I ran Cactus directly on the central manager node), I had a look at the worker nodes and while lastz was running, I noticed that only around 20-25% of the CPU resources were actually in consistent use. Any idea on what the reason for this might be?

**Comment by amizeranschi** [comment]
My best guess so far for the relatively low CPU usage is that disk I/O is causing slow-downs. The cluster has NFS storage and the jobstore is set up there. I've read about caching in Toil being able to decrease the impact of I/O: https://toil.readthedocs.io/en/3.14.0/appendices/architecture.html#caching

However, I'm stuck with running Cactus with `--disableCaching`, because this option doesn't seem to be supported when using htcondor. See: https://github.com/ComparativeGenomicsToolkit/cactus/issues/903#issuecomment-1387175592

Any advice on how I could optimize the CPU usage / decrease I/O of Cactus on this htcondor cluster?

**Comment by adamnovak** [comment]
For optimizing Cactus's resource use you might want to talk to @glennhickey on an issue in the Cactus repo. My best guess would be either that it is indeed IO bound, that lastz just doesn't need that much CPU (in which case you could lower the amount of CPU that Cactus requests for those jobs to reduce waste), or that *some* lastz jobs need more CPU than others and Cactus is applying just one resource request size to all of them because it doesn't know how to tell them apart.

The Toil caching system won't reduce NFS IO that much unless your NFS doesn't do its own caching or your jobs are relatively short. Without caching, each job will copy all the files it uses fresh from the job store (here on NFS) to local disk, before using them. If NFS/the kernel does its own caching that could be reasonably fast, and it shouldn't show up in CPU usage of the lastz processes themselves, because they won't be reading from NFS but rather from local disk. Instead, you would see each job running single-threaded for some amount of time waiting for the copy to go through, before starting lastz.

Right now we don't let you turn on the caching on HTCondor because we don't get an opportunity to clean up each node after the last job.

I feel like we could engineer a solution for the caching that doesn't depend on cleaning up each node after the last job, but it would require a way to have multiple caches and a way for a newly starting job to connect to the right one, because we need to handle the case where the last job running on a node clears out the cache, and then while it is doing that a new job arrives and wants to use the cache.

But depending on the job pattern in the workflow and how htcondor schedules, that might not end up being very efficient. If at any point a node isn't running a Toil job, the cache for that node would have to be fully flushed, and the next job to start on that node wouldn't be able to benefit from it.

Maybe we need a bypass flag to let people run caching even when it will leave abandoned temporary files taking up unshared local disk on the execution nodes?

If we did abandon cached files on the local disks of your clusters' nodes, would you somehow be able to schedule a cleanup job to execute on *every* node in the cluster, without using an inordinate amount of your cluster's resources?

**Comment by amizeranschi** [comment]
@adamnovak and @glennhickey 

Seems like I was too quick to (negatively) judge. I ran some tests on a specific scenario, involving 2 plant genomes: Arabidopsis Thaliana TAIR10 from Ensembl and another plant genome, around 800 Mbp in length. I ran three tests for this scenario, as such:

1. the first one via Docker, locally, on a local file system (XFS), using 128 AMD Epyc cores, with around 7.5 GB per core.
2. the second one also via Docker, locally, but on NFS storage, also with 128 AMD Epyc cores, with around 7.5 GB per core.
3. the third one with Cactus and Toil installed in virtualenv, run on an HTCondor cluster (with NFS storage), as mentioned here: https://github.com/DataBiosphere/toil/pull/4396#issuecomment-1458806381. The execution nodes were 2 x 64 AMD Epyic cores, with around 7.5 GB per core. 

In all cases, the Cactus version was 2.4.1. The runtimes for the three scenarios, in order of them being mentioned, were as such:

```
[2023-02-23T17:33:26+0000] [MainThread] [I] [toil.statsAndLogging] Cactus has finished after 15416.366406648885 seconds
[2023-03-16T22:14:29+0000] [MainThread] [I] [toil.statsAndLogging] Cactus has finished after 16227.541185555048 seconds
[2023-03-17T17:50:27+0100] [MainThread] [I] [toil.statsAndLogging] Cactus has finished after 16374.566635811003 seconds
```

So, it looks like both NFS and HTCondor had a relatively low impact on performance, in these scenarios. I'm honestly very happy with this outcome and will be considering the HTCondor + NFS scenario for more complex jobs.

For the record, the native HTCondor version was 8.8.15 and the Python htcondor module version was 8.9.13.

Our plan for the next few weeks is to upgrade to htcondor 10.X and see how well things work out in that case. I can update here on the outcome, in case it helps.
